File: client/cli/src/engine.rs
//! Engine client utilities for the Pie CLI.

use crate::path;
use anyhow::{Context, Result};
use pie_client::client::{Client, Instance, InstanceEvent};
use pie_client::crypto::ParsedPrivateKey;
use pie_client::message::EventCode;
use std::fs;
use std::path::PathBuf;
use tokio::signal;

// Helper struct for what client commands need to know
pub struct ClientConfig {
    pub host: String,
    pub port: u16,
    pub username: String,
    pub private_key: Option<ParsedPrivateKey>,
    pub enable_auth: bool,
}

impl ClientConfig {
    pub fn new(
        config_path: Option<PathBuf>,
        host: Option<String>,
        port: Option<u16>,
        username: Option<String>,
        private_key_path: Option<PathBuf>,
    ) -> Result<Self> {
        // Read config file only if when any parameter is missing
        let config_file =
            if host.is_none() || port.is_none() || username.is_none() || private_key_path.is_none()
            {
                let config_str = match config_path {
                Some(path) => fs::read_to_string(&path)
                    .context(format!("Failed to read config file at {:?}", path))?,
                None => fs::read_to_string(&crate::path::get_default_config_path()?).context(
                    "Failed to read default config file. Try running `pie-cli config init` first.",
                )?,
            };
                Some(toml::from_str::<crate::config::ConfigFile>(&config_str)?)
            } else {
                None
            };

        // Prefer command-line arguments and use config file values if not provided
        let host = host
            .or_else(|| config_file.as_ref().and_then(|cfg| cfg.host.clone()))
            .unwrap_or("127.0.0.1".to_string());

        let port = port
            .or_else(|| config_file.as_ref().and_then(|cfg| cfg.port))
            .unwrap_or(8080);

        let username = username
            .or(config_file.as_ref().and_then(|cfg| cfg.username.clone()))
            .unwrap_or(whoami::username());

        // Get enable_auth setting (default to true if not specified)
        let enable_auth = config_file
            .as_ref()
            .and_then(|cfg| cfg.enable_auth)
            .unwrap_or(true);

        let mut private_key = None;

        // Load the private key only if authentication is enabled
        if enable_auth {
            let private_key_path = private_key_path
                .or(config_file
                    .as_ref()
                    .and_then(|cfg| cfg.private_key_path.clone()))
                .map(|p| {
                    p.to_str()
                        .map(|s| s.to_owned())
                        .context("Private key path is not a valid UTF-8 string")
                })
                .transpose()?
                .map(|p| path::expand_tilde(&p))
                .transpose()?
                .context("Private key is required when authentication is enabled")?;

            // Check private key file permissions (Unix only)
            #[cfg(unix)]
            path::check_private_key_permissions(&private_key_path)?;

            // Read and parse the private key from the file if a path is provided
            let key_content = fs::read_to_string(&private_key_path).context(format!(
                "Failed to read private key file at {:?}",
                private_key_path
            ))?;
            private_key = Some(ParsedPrivateKey::parse(&key_content).context(format!(
                "Failed to parse private key from {:?}",
                private_key_path
            ))?);
        }

        Ok(Self {
            host,
            port,
            username,
            private_key,
            enable_auth,
        })
    }
}

/// Streams the output of an inferlet.
///
/// Behavior:
/// - Ctrl-C (SIGINT): Sends a terminate request to the server to kill the inferlet
/// - Ctrl-D (EOF): Detaches from the inferlet (continues running on server)
pub async fn stream_inferlet_output(mut instance: Instance, client: Client) -> Result<()> {
    let instance_id = instance.id().to_string();
    let short_id = instance_id[..instance_id.len().min(8)].to_string();
    let mut at_line_start_stdout = true;
    let mut at_line_start_stderr = true;

    // Set up Ctrl-C signal handler
    let mut sigint = signal::unix::signal(signal::unix::SignalKind::interrupt())
        .context("Failed to set up SIGINT handler")?;

    // We use a separate OS thread with blocking I/O to monitor stdin for EOF, rather than
    // using `tokio::io::stdin()`. This is because async stdin puts the terminal in non-blocking
    // mode, which can leave the terminal in an inconsistent state if we exit through other
    // paths (e.g., when the inferlet completes naturally). The blocking thread approach keeps
    // the terminal in normal mode and ensures clean process termination.
    let (eof_tx, mut eof_rx) = tokio::sync::mpsc::channel::<()>(1);
    let _stdin_monitor = std::thread::spawn(move || {
        use std::io::{self, Read};
        let mut stdin = io::stdin();
        let mut buf = [0u8; 1];

        loop {
            match stdin.read(&mut buf) {
                Ok(0) | Err(_) => {
                    // EOF detected (Ctrl-D pressed) or stdin closed
                    let _ = eof_tx.blocking_send(());
                    break;
                }
                Ok(_) => {
                    // Ignore any input data and continue monitoring
                    continue;
                }
            }
        }
    });

    // Main event loop: stream output and handle signals
    loop {
        tokio::select! {
            // Ctrl-C: Send termination request to server
            _ = sigint.recv() => {
                println!("\n[Instance {}] Received Ctrl-C, terminating instance ...", short_id);
                if let Err(e) = client.terminate_instance(&instance_id).await {
                    eprintln!("[Instance {}] Failed to send terminate request: {}", short_id, e);
                }
                return Ok(());
            }

            // Ctrl-D: Detach without terminating
            _ = eof_rx.recv() => {
                println!("\n[Instance {}] Detached from instance ...", short_id);
                return Ok(());
            }

            // Process instance events (output, completion, errors)
            event_result = instance.recv() => {
                let event = match event_result {
                    Ok(ev) => ev,
                    Err(e) => {
                        println!("[Instance {}] ReceiveError: {}", short_id, e);
                        return Err(e);
                    }
                };

                match event {
                    InstanceEvent::Event { code, message } => {
                        println!("[Instance {}] {:?}: {}", short_id, code, message);

                        match code {
                            EventCode::Completed => return Ok(()),
                            EventCode::Message => continue,
                            EventCode::Aborted
                            | EventCode::Exception
                            | EventCode::ServerError
                            | EventCode::OutOfResources => {
                                anyhow::bail!("inferlet terminated with status {:?}", code)
                            }
                        }
                    }

                    InstanceEvent::Stdout(content) => {
                        handle_streaming_output(false, content, &short_id, &mut at_line_start_stdout).await;
                    }

                    InstanceEvent::Stderr(content) => {
                        handle_streaming_output(true, content, &short_id, &mut at_line_start_stderr).await;
                    }

                    InstanceEvent::Blob(_) => {
                        // Ignore binary blobs
                        continue;
                    }
                }
            }
        }
    }
}

/// Helper to handle streaming output (stdout or stderr).
async fn handle_streaming_output(
    is_stderr: bool,
    content: String,
    short_id: &str,
    at_line_start: &mut bool,
) {
    let at_start = *at_line_start;
    let short_id = short_id.to_string();
    *at_line_start = tokio::task::spawn_blocking(move || {
        let mut at_start_local = at_start;
        if is_stderr {
            write_with_prefix(
                std::io::stderr().lock(),
                &content,
                &short_id,
                &mut at_start_local,
            );
        } else {
            write_with_prefix(
                std::io::stdout().lock(),
                &content,
                &short_id,
                &mut at_start_local,
            );
        }
        at_start_local
    })
    .await
    .unwrap_or(at_start);
}

/// Helper function to print output with instance ID prefix only at the start of new lines.
fn write_with_prefix(
    mut writer: impl std::io::Write,
    content: &str,
    short_id: &str,
    at_line_start: &mut bool,
) {
    if content.is_empty() {
        return;
    }

    let lines = content.split('\n');
    let mut first = true;

    for line in lines {
        if !first {
            // We encountered a '\n' separator, print it
            let _ = writeln!(writer);
            *at_line_start = true;
        }
        first = false;

        // Add prefix only if we're at line start and the line is non-empty
        if !line.is_empty() {
            if *at_line_start {
                let _ = write!(writer, "[Instance {}] ", short_id);
                *at_line_start = false;
            }
            let _ = write!(writer, "{}", line);
        }
    }
}

/// Connects to the engine and authenticates the client if authentication is enabled.
pub async fn connect_and_authenticate(client_config: &ClientConfig) -> Result<Client> {
    let url = format!("ws://{}:{}", client_config.host, client_config.port);
    let client = match Client::connect(&url).await {
        Ok(c) => c,
        Err(_) => {
            anyhow::bail!("Could not connect to engine at {}. Is it running?", url);
        }
    };

    let result = client
        .authenticate(&client_config.username, &client_config.private_key)
        .await;

    if client_config.enable_auth {
        result.context("Failed to authenticate with engine using the specified private key")?;
    } else {
        result.context(
            "Failed to authenticate with engine (client public key authentication disabled)",
        )?;
    }

    Ok(client)
}

File: client/python/src/pie_client_cli/engine.py
"""Engine client utilities for the Pie CLI.

This module provides utilities for connecting to the Pie engine, authentication,
and streaming inferlet output with signal handling.
"""

import asyncio
import os
import signal
import sys
import threading
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

import typer

from pie_client import PieClient, Event, Instance, InstanceInfo, LibraryInfo
from pie_client.crypto import ParsedPrivateKey

from . import path as path_utils
from .config import ConfigFile


@dataclass
class ClientConfig:
    """Configuration for connecting to the Pie engine."""

    host: str
    port: int
    username: str
    private_key: Optional[ParsedPrivateKey]
    enable_auth: bool

    @classmethod
    def create(
        cls,
        config_path: Optional[Path] = None,
        host: Optional[str] = None,
        port: Optional[int] = None,
        username: Optional[str] = None,
        private_key_path: Optional[Path] = None,
    ) -> "ClientConfig":
        """Create a ClientConfig, merging command-line args with config file.

        Command-line arguments take precedence over config file values.
        """
        # Read config file if any parameter is missing
        config_file: Optional[ConfigFile] = None
        if host is None or port is None or username is None or private_key_path is None:
            config_file_path = config_path or path_utils.get_default_config_path()
            if config_file_path.exists():
                config_file = ConfigFile.load(config_file_path)
            elif config_path is not None:
                # User explicitly specified a config path that doesn't exist
                raise FileNotFoundError(f"Config file not found at '{config_path}'")
            # If default config doesn't exist, we'll use defaults below

        # Merge with defaults
        final_host = host or (config_file.host if config_file else None) or "127.0.0.1"
        final_port = port or (config_file.port if config_file else None) or 8080
        final_username = (
            username or (config_file.username if config_file else None) or os.getlogin()
        )

        # Get enable_auth setting
        enable_auth = (
            config_file.enable_auth
            if config_file and config_file.enable_auth is not None
            else True
        )

        # Load private key if auth is enabled
        private_key: Optional[ParsedPrivateKey] = None
        if enable_auth:
            key_path = private_key_path
            if key_path is None and config_file and config_file.private_key_path:
                key_path = Path(config_file.private_key_path).expanduser()

            if key_path is None:
                raise ValueError(
                    "Private key is required when authentication is enabled. "
                    "Set private_key_path in config or use --private-key-path."
                )

            key_path = key_path.expanduser()

            # Check permissions on Unix
            if os.name == "posix":
                path_utils.check_private_key_permissions(key_path)

            # Read and parse the private key
            key_content = key_path.read_text()
            private_key = ParsedPrivateKey.parse(key_content)

        return cls(
            host=final_host,
            port=final_port,
            username=final_username,
            private_key=private_key,
            enable_auth=enable_auth,
        )


async def _connect_and_authenticate_async(client_config: ClientConfig) -> PieClient:
    """Connect to the engine and authenticate (async version)."""
    url = f"ws://{client_config.host}:{client_config.port}"

    client = PieClient(url)
    try:
        await client.connect()
    except Exception:
        raise ConnectionError(f"Could not connect to engine at {url}. Is it running?")

    try:
        await client.authenticate(client_config.username, client_config.private_key)
    except Exception as e:
        await client.close()
        if client_config.enable_auth:
            raise ConnectionError(
                f"Failed to authenticate with engine using the specified private key: {e}"
            )
        else:
            raise ConnectionError(
                f"Failed to authenticate with engine (client public key authentication disabled): {e}"
            )

    return client


def connect_and_authenticate(client_config: ClientConfig) -> PieClient:
    """Connect to the engine and authenticate (sync wrapper)."""
    return asyncio.get_event_loop().run_until_complete(
        _connect_and_authenticate_async(client_config)
    )


def _write_with_prefix(
    is_stderr: bool,
    content: str,
    short_id: str,
    at_line_start: bool,
) -> bool:
    """Write output with instance ID prefix at line starts.

    Returns the new at_line_start state.
    """
    if not content:
        return at_line_start

    writer = sys.stderr if is_stderr else sys.stdout
    lines = content.split("\n")
    first = True

    for line in lines:
        if not first:
            # We encountered a '\n' separator
            writer.write("\n")
            at_line_start = True
        first = False

        # Add prefix only if at line start and line is non-empty
        if line:
            if at_line_start:
                writer.write(f"[Instance {short_id}] ")
                at_line_start = False
            writer.write(line)

    writer.flush()
    return at_line_start


async def _stream_inferlet_output_async(
    instance: Instance,
    client: PieClient,
) -> None:
    """Stream output from an inferlet with signal handling (async version).

    Behavior:
    - Ctrl-C (SIGINT): Sends terminate request to the server
    - Ctrl-D (EOF on stdin): Detaches from the inferlet (continues running)
    """
    instance_id = instance.instance_id
    short_id = instance_id[: min(8, len(instance_id))]
    at_line_start_stdout = True
    at_line_start_stderr = True

    # Set up SIGINT handling
    sigint_received = asyncio.Event()
    original_handler = signal.getsignal(signal.SIGINT)

    def sigint_handler(signum, frame):
        sigint_received.set()

    signal.signal(signal.SIGINT, sigint_handler)

    # Set up stdin EOF detection in a separate thread
    eof_received = asyncio.Event()

    def stdin_monitor():
        try:
            while True:
                data = sys.stdin.read(1)
                if not data:  # EOF
                    asyncio.get_event_loop().call_soon_threadsafe(eof_received.set)
                    break
        except Exception:
            pass

    stdin_thread = threading.Thread(target=stdin_monitor, daemon=True)
    stdin_thread.start()

    try:
        while True:
            # Create tasks for the three events we're waiting on
            recv_task = asyncio.create_task(instance.recv())
            sigint_wait = asyncio.create_task(sigint_received.wait())
            eof_wait = asyncio.create_task(eof_received.wait())

            done, pending = await asyncio.wait(
                [recv_task, sigint_wait, eof_wait],
                return_when=asyncio.FIRST_COMPLETED,
            )

            # Cancel pending tasks
            for task in pending:
                task.cancel()
                try:
                    await task
                except asyncio.CancelledError:
                    pass

            # Handle SIGINT (Ctrl-C)
            if sigint_wait in done:
                typer.echo(
                    f"\n[Instance {short_id}] Received Ctrl-C, terminating instance ..."
                )
                try:
                    await client.terminate_instance(instance_id)
                except Exception as e:
                    typer.echo(
                        f"[Instance {short_id}] Failed to send terminate request: {e}",
                        err=True,
                    )
                return

            # Handle EOF (Ctrl-D)
            if eof_wait in done:
                typer.echo(f"\n[Instance {short_id}] Detached from instance ...")
                return

            # Handle instance event
            if recv_task in done:
                try:
                    event, message = recv_task.result()
                except Exception as e:
                    print(f"[Instance {short_id}] ReceiveError: {e}")
                    raise

                if event == Event.Message:
                    typer.echo(f"[Instance {short_id}] Message: {message}")

                elif event == Event.Completed:
                    typer.echo(f"[Instance {short_id}] Completed: {message}")
                    return

                elif event in (
                    Event.Aborted,
                    Event.Exception,
                    Event.ServerError,
                    Event.OutOfResources,
                ):
                    typer.echo(f"[Instance {short_id}] {event.name}: {message}")
                    raise RuntimeError(f"inferlet terminated with status {event.name}")

                elif event == Event.Stdout:
                    at_line_start_stdout = _write_with_prefix(
                        False, message, short_id, at_line_start_stdout
                    )

                elif event == Event.Stderr:
                    at_line_start_stderr = _write_with_prefix(
                        True, message, short_id, at_line_start_stderr
                    )

                elif event == Event.Blob:
                    # Ignore binary blobs
                    pass

    finally:
        # Restore original signal handler
        signal.signal(signal.SIGINT, original_handler)


def stream_inferlet_output(instance: Instance, client: PieClient) -> None:
    """Stream output from an inferlet with signal handling (sync wrapper)."""
    asyncio.get_event_loop().run_until_complete(
        _stream_inferlet_output_async(instance, client)
    )


# Sync wrappers for common client operations
def ping(client: PieClient) -> None:
    """Ping the server (sync wrapper)."""
    asyncio.get_event_loop().run_until_complete(client.ping())


def list_instances(client: PieClient) -> list[InstanceInfo]:
    """List instances (sync wrapper)."""
    return asyncio.get_event_loop().run_until_complete(client.list_instances())


def terminate_instance(client: PieClient, instance_id: str) -> None:
    """Terminate an instance (sync wrapper)."""
    asyncio.get_event_loop().run_until_complete(client.terminate_instance(instance_id))


def attach_instance(client: PieClient, instance_id: str) -> Instance:
    """Attach to an instance (sync wrapper)."""
    return asyncio.get_event_loop().run_until_complete(
        client.attach_instance(instance_id)
    )


def upload_program(client: PieClient, program_bytes: bytes) -> None:
    """Upload a program (sync wrapper)."""
    asyncio.get_event_loop().run_until_complete(client.upload_program(program_bytes))


def program_exists(client: PieClient, program_hash: str) -> bool:
    """Check if a program exists (sync wrapper)."""
    return asyncio.get_event_loop().run_until_complete(
        client.program_exists(program_hash)
    )


def launch_instance(
    client: PieClient,
    program_hash: str,
    arguments: list[str],
    detached: bool = False,
) -> Instance:
    """Launch an instance (sync wrapper)."""
    return asyncio.get_event_loop().run_until_complete(
        client.launch_instance(program_hash, arguments, detached)
    )


def launch_instance_from_registry(
    client: PieClient,
    inferlet: str,
    arguments: list[str],
    detached: bool = False,
) -> Instance:
    """Launch an instance from the registry (sync wrapper).

    The inferlet parameter can be:
    - Full name with version: "std/text-completion@0.1.0"
    - Without namespace (defaults to "std"): "text-completion@0.1.0"
    - Without version (defaults to "latest"): "std/text-completion" or "text-completion"
    """
    return asyncio.get_event_loop().run_until_complete(
        client.launch_instance_from_registry(inferlet, arguments, detached)
    )


def close_client(client: PieClient) -> None:
    """Close the client (sync wrapper)."""
    asyncio.get_event_loop().run_until_complete(client.close())


def upload_library(
    client: PieClient,
    name: str,
    library_bytes: bytes,
    dependencies: list[str] | None = None,
) -> None:
    """Upload a library (sync wrapper)."""
    asyncio.get_event_loop().run_until_complete(
        client.upload_library(name, library_bytes, dependencies)
    )


def list_libraries(client: PieClient) -> list[LibraryInfo]:
    """List loaded libraries (sync wrapper)."""
    return asyncio.get_event_loop().run_until_complete(client.list_libraries())
